File: 백준/Silver/2468. 안전 영역/안전 영역.cc
#include <bits/stdc++.h>

using namespace std;

const int dy[] = { -1,0,1,0 };
const int dx[] = { 0,1,0,-1 };

int a[102][102], visited[102][102];
int N, H, ret;
int mret = -100;
int m = 100;

void DFS(int y, int x) {
	visited[y][x] = 1;
	for (int i = 0; i < 4; i++) {
		int ny = y + dy[i];
		int nx = x + dx[i];
		if (ny < 0 || nx < 0 || ny >= N || nx >= N) continue;
		if (!visited[ny][nx] && a[ny][nx] >= H) DFS(ny, nx);
	}
	return;
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL); cout.tie(NULL);
	
	cin >> N;

	for (int i = 0; i < N; i++) {
		for (int j = 0; j < N; j++) {
			cin >> a[i][j];
		}
	}
	while (m--) {
		memset(visited, 0, sizeof(visited));
		ret = 0;
		for (int i = 0; i < N; i++) {
			for (int j = 0; j < N; j++) {
				if (!visited[i][j] && a[i][j] >= H) {
					ret++; DFS(i, j);
				}
			}
		}
		mret = max(ret, mret);
		H++;
	}
	cout << mret << '\n';
	return 0;
}

File: 백준/Silver/2468. 안전 영역/안전 영역.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    static int[][] maps;
    static int[][] visited;
    static int[] dy = {-1, 0, 1, 0};
    static int[] dx = {0, 1, 0, -1};

    static void dfs(int y, int x, int water) {
        visited[y][x] = 1;
        for (int i = 0; i < 4; i++) {
            int ny = y + dy[i];
            int nx = x + dx[i];

            if (ny < 0 || nx < 0 || ny >= maps.length || nx >= maps[0].length || visited[ny][nx] == 1) {
                continue;
            }
            if (maps[ny][nx] > water) {
                dfs(ny, nx, water);
            }
        }
    }


    public static void main(String[] args) throws IOException {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(bf.readLine());
        maps = new int[N][N];
        int max = 0;
        for (int i = 0; i < N; i++) {
            StringTokenizer st = new StringTokenizer(bf.readLine());
            for (int j = 0; j < N; j++) {
                int height = Integer.parseInt(st.nextToken());
                max = Math.max(max, height);
                maps[i][j] = height;
            }
        }
        int water = 0;
        int ret = 0;
        while (water <= max) {
            visited = new int[N][N];
            int safe = 0;
            for (int i = 0; i < N; i++) {
                for (int j = 0; j < N; j++) {
                    if (maps[i][j] > water && visited[i][j] == 0) {
                        dfs(i, j, water);
                        safe++;
                    }
                }
            }
            ret = Math.max(ret, safe);
            water++;
        }
        System.out.println(ret);
    }
}